### PR TITLE
WPWebViewController: Revamp Mark II

### LINKED ISSUE #3763: WPWebViewController: Revamp Mark II
After some testing, we've realized that it's not a good idea to always fall back to *Safari*, whenever the user taps over an external link (Reference: #3752).

Let's review every single flow in which WebViewController is displayed, and:

- [x] Convert the transition to Modal, when applicable.
- [x] Update the NavigationBar Style, when displayed modally.
- [x] Nuke the Pull to Refresh control.
- [x] Remove the 'Open in Safari' AlertView.
- [x] Bring back the bottom toolbar (as per @drw158's awesome design).
- [x] Bottom Toolbar should be hidden, whenever both Back + Forward are disabled.

--

#### Mockup:

iphone_6

### PR DESCRIPTION
**Pull request by jleandroperez**
#### Description:
This Pull Request revisits all of the current interactions with WPWebViewController, and converts (almost all of them) into Modal interactions. Exceptions outlined below.

Also, several UI enhancements have been introduced: the bottom toolbar has been brought back, and whenever the WebViewController is displayed modally, it'll skin the NavigationBar with a different style.

#### Notes:
- **StatsWebViewController** _might_ push a **WebViewController** instance, when the user taps over an external link. We'll deal with those interactions as we always did [as a navigationController push].
- **AboutViewController** can display  **WebViewController** instances. Since About can be displayed either modally or "pushed". We'll avoid presenting the webViewController modally, if we're already in a modal scenario.
- Post previews are actually handled by **PostPreviewViewController**, so this PR doesn't affect that functionality.

#### Review:
Needs Review: @aerych + @astralbodies 

Eric + Aaron: may i bother you guys with a review?. Code delta isn't huge, but it'll probably take a while to test everything. Maybe 7 test cases / 7 test cases, and raffle the last one?

(Thanks in advance!!)

Closes #3763

--

#### Test 1: Login
1. Fresh install WPiOS
2. Tap over the 'Add Self-Hosted Site' label
3. Enter invalid credentials, and hit Sign In
4. Once you've gotthe 'Sorry, we can't log you in' screen, tap over the bottom left icon that reads 'Need Help?'

**Expected:** Modal WebViewController onscreen

--

#### Test 2: Create Account
1. Launch WPiOS
2. Tap over Create Account
3. Tap the Terms of Service label

**Expected:** Modal WebViewController onscreen

--

#### Test 3: About (Logged Off)
1. Fresh install WPiOS
2. Tap over the top right (?) icon
3. Tap over About, and pick any of the links

**Expected:** WebViewController onscreen, with the regular NavBar style

--

#### Test 4: About (Logged In)
1. Log into WPiOS
2. Tap over the Me tab, and hit the *Help & Support* row
3. Tap over About, and pick any of the links

**Expected:** Modal WebViewController onscreen

--

#### Test 5: View Site
1. Launch WPiOS and log into your account
2. Tap over My Sites, and pick a random blog
3. Tap over the *View Site* label

**Expected:** Modal WebViewController onscreen

--

#### Test 6: Reader Post Image
1. Launch WPiOS and log into your account
2. Tap over the Reader and pick any article
3. Tap over an image that's not PNG | JPG | GIF

**Expected:** Modal WebViewController onscreen
**Note:** This is way easier to test if you just hack *WPImageViewController*'s isUrlSupported method, to always return NO.

--

#### Test 7: Reader Comment Image
1. Launch WPiOS and log into your account
2. Tap over the Reader, pick any article, and open its comments
3. Tap over an image that's not PNG | JPG | GIF

**Expected:** Modal WebViewController onscreen
**Note:* This is way easier to test if you just hack *WPImageViewController*'s isUrlSupported method, to always return NO.

--

#### Test 8: Reader Post Link
1. Launch WPiOS and log into your account
2. Tap over the Reader and pick any article
3. Tap over any external links

**Expected:** Modal WebViewController onscreen

--

#### Test 9: Reader Comment Link
1. Launch WPiOS and log into your account
2. Tap over the Reader, pick any article, and open its comments
3. Tap over any external links

**Expected:** Modal WebViewController onscreen

--

#### Test 10: Reader Browse Site
1. Launch WPiOS and log into your account
2. Tap over the Reader, and tap over the top left icon of any article (Blavatar)
3. Tap again over the Blavatar

**Expected:** Modal WebViewController onscreen

--

#### Test 11: Notifications Jetpack
1. Launch WPiOS and log into a self hosted blog, with no Jetpack
2. Tap over the Notifications tab
3. Tap over the "Learn More" button

**Expected:** Modal WebViewController onscreen
**Note:** This is easier to test with a WordPress.com account with no notifications. Just hack the *NotificationsViewController*'s showsJetpackMessage method to always return YES.

--

#### Test 12: Notification Details
1. Launch WPiOS and log into your account
2. Tap over the Notifications tab, and tap over any notification that contains a web link (ie. badge / comment)
3. Tap over the link

**Expected:** Modal WebViewController onscreen

--

#### Test 13: Comments
1. Launch WPiOS and log into your account
2. Tap over My Sites, and pick a random blog
3. Tap over 'Comments', and pick a comment that contains a web link
4. Tap over the link

**Expected:** Modal WebViewController onscreen

--

#### Test 14: Edit Site
1. Log into a self hosted blog
2. Tap over My Sites, and pick up the blog
3. Force an error 405 (easiest way: hack the *validationDidFail* method in EditSiteViewController, change the password onscreen, and hit save).

**Expected:** Modal WebViewController onscreen

-- 

#### Test 15: Install Jetpack
1. Log into a self hosted blog, that doesn't have Jetpack installed
2. Tap over My Sites > Stats
3. Tap over the 'Install Jetpack' button

**Expected:** Modal WebViewController onscreen

### CODE REVIEW

File: WordPress/Classes/Utility/WPStyleGuide+WebView.m [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+@implementation WPStyleGuide (WebView)
+
++ (UIColor *)webViewModalNavigationBarBackground
+{
+    return [UIColor whiteColor];
+}
+
++ (UIColor *)webViewModalNavigationBarShadow
+{
+    return [UIColor colorWithRed:(CGFloat)46/255.0 green:(CGFloat)68/255.0 blue:(CGFloat)83/255.0 alpha:0.15];
```
Comment (aerych) [inline WordPress/Classes/Utility/WPStyleGuide+WebView.m]: Why the cast to a CGFloat? Maybe just add a .0 to the numerator?

Comment (jleandroperez) [inline WordPress/Classes/Utility/WPStyleGuide+WebView.m]: Ouch! that's the result of a quick copy + paste. Fixing, thanks!

File: WordPress/Classes/ViewRelated/Reader/ReaderCommentsViewController.m [hunk 2]
```diff
@@ -1142,42 +1143,31 @@ - (void)richTextView:(WPRichTextView *)richTextView didReceiveLinkAction:(NSURL
...
-                break;
-            }
-        }
-
-        if (matched) {
-            controller = [[WPImageViewController alloc] initWithImage:imageControl.imageView.image andURL:imageControl.linkURL];
-        } else {
-            controller = [WPWebViewController webViewControllerWithURL:imageControl.linkURL];
-        }
+    UIViewController *controller = nil;
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Reader/ReaderCommentsViewController.m]: An object variable declaration without an assignment is automatically assigned nil by the compiler.  Yay for llvm.

**Comment by aerych** [comment]
Looks like descenders can extend below the top bar in landscape mode.

ios simulator screen shot jun 3 2015 1 08 47 pm

**Comment by aerych** [comment]
I get crashes on the iPad when testing:
> 2015-06-03 13:53:40.391 WordPress[10960:105132] *** Terminating app due to uncaught exception 'NSUnknownKeyException', reason: '[<WPWebViewController 0x79ebe750> setValue:forUndefinedKey:]: this class is not key value coding-compliant for the key activityIndicator.'

**Comment by aerych** [comment]
I ran through all the tests on the iPhone except for 2 and 14 and everything passed.  We need to follow up with a round on the iPad when its happy.

**Comment by aerych** [comment]
Oopsie. The iPad version is also working fine after cleaning. Blame it on a stale cached nib.

File: WordPress/Classes/ViewRelated/Reader/ReaderCommentsViewController.m (hunk 2 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Reader/ReaderCommentsViewController.m]: Old habits die hard! but it's good to know we're safe anyways!

**Comment by jleandroperez** [comment]
@aerych descender glitch fixed (moved everything 1 point up). Nice catch there!.

Please, let me know if i can help with scenarios 2 / 14 (got a couple self hosted blogs).

*Thanks* Eric!

**Comment by aerych** [comment]
Scenario 2 passes. Not sure why I skipped it before. 
For Scenario 14, what was the "hack" you had in mind? I'm missing it.

**Comment by aerych** [comment]
If you look at the earlier screen shot of the web view showing the Terms of Service page you can see there is a lot of room in the navbar, but the title is still being cropped.  
Maybe we could edit `NavigationTitleView.setupSubviews` and include a `UIViewAutoresizing.FlexibleWidth` mask so we can use more of the available space?  Like so:

ios simulator screen shot jun 4 2015 10 08 43 am

**Comment by aerych** [comment]
Tested scenario 14 (the last of the batch) and it succeeds.

**Comment by aerych** [comment]
@jleandroperez I think once the title size is addressed its good to :shipit: :)

**Comment by jleandroperez** [comment]
@aerych NavigationTitleView fix implemented, exactly as you mentioned!. Thank you Eric!

**Pull request merged**